File: leetcode/680. Valid Palindrome II.cpp
class Solution {
public:
    bool validPalindrome(string s) {
        int i = 0;
        int j = s.size() - 1;
        while(i < j){
            if(s[i] != s[j]){
                return isValid(s, i + 1, j) || isValid(s, i, j - 1);
            }
            ++i;
            --j;
        }
        return true;
    }
private:
    bool isValid(string& s, int i, int j) {
        while(i < j){
            if(s[i++] != s[j--]){
                return false;
            }
        }
        return true;
    }
};

File: leetcode/680. Valid Palindrome II.rs
impl Solution {
    pub fn valid_palindrome(s: String) -> bool {
        let v: Vec<char> = s.chars().collect();
        let mut i = 0;
        let mut j = v.len() - 1;
        while i < j {
            if v[i] != v[j] {
                return Solution::is_palindrome(&v, i, j - 1) || Solution::is_palindrome(&v, i + 1, j);
            }
            i += 1;
            j -= 1;
        }
        return true;
    }
    
    fn is_palindrome(v: &Vec<char>, mut i: usize , mut j: usize) -> bool {
        while i < j {
            if v[i] != v[j] {
                return false;
            }
            i += 1;
            j -= 1;
        }
        return true;
    }
}
